tests/python/test_multidevice.py

```diff
@@ -332,35 +332,49 @@ def definition(self) -> None:
         T78 = self.ops.linear(T77, self.mha_linear0_weight, self.mha_linear0_bias)
         T91 = self.ops.slice(
             T78,
-            start_indices=[0, 0, 0],
-            end_indices=[b, s, e],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, 0],
+            end_indices=[d, b, s, e // d],
         )
         T104 = self.ops.slice(
             T78,
-            start_indices=[0, 0, e],
-            end_indices=[b, s, e * 2],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, e // d],
+            end_indices=[d, b, s, e * 2 // d],
         )
         T117 = self.ops.slice(
             T78,
-            start_indices=[0, 0, e * 2],
-            end_indices=[b, s, e * 3],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, e * 2 // d],
+            end_indices=[d, b, s, e * 3 // d],
         )
-        T123 = self.ops.reshape(T104, new_shape=[b, s, h, e // h])
-        T124 = self.ops.permute(T123, dims=[0, 2, 1, 3])
-        T130 = self.ops.reshape(T91, new_shape=[b, s, h, e // h])
-        T131 = self.ops.permute(T130, dims=[0, 2, 1, 3])
-        T137 = self.ops.reshape(T117, new_shape=[b, s, h, e // h])
-        T138 = self.ops.permute(T137, dims=[0, 2, 1, 3])
+        T123 = self.ops.reshape(T104, new_shape=[d, b, s, h // d, e // h])
```

Comment:
IIUC, we pass from shape `[d, b, s, e//d]` to `[d, b, s, h//d, e//h]`. Nothing illegal about it but it looks surprising to me so I just want to make sure

Reply:
I double checked -- it looks right. MHA is head parallel according to Figure 3b in https://arxiv.org/pdf/1909.08053.